nixos/modules/services/networking/seafile.nix

```diff
@@ -0,0 +1,272 @@
+      default = "";
+      type = types.lines;
+      description = ''
+        Extra config to append to `seafile.conf` file.
+        This is appended under the fileserver section.
+        Refer to <link xlink:href="https://manual.seafile.com/config/seafile-conf/" />
+        for all available options.
+      '';
+    };
+
+    adminEmail = mkOption {
+      example = "john@example.com";
+      type = types.str;
+      description = ''
+        Seafile Seahub Admin Account Email.
+      '';
+    };
+
+    initialAdminPassword = mkOption {
+      example = "someStrongPass";
+      type = types.str;
+      description = ''
+        Seafile Seahub Admin Account initial password.
+        Should be change via Seahub web front-end.
+      '';
+    };
+
+    seahubSecretKeyFile = mkOption {
+      type = types.path;
+      description = ''
+        A file containing the django secret key for safe session.
+      '';
+    };
+
+    seahubExtraConf = mkOption {
+      default = "";
+      type = types.lines;
+      description = ''
+        Extra config to append to `seahub_settings.py` file.
+        Refer to <link xlink:href="https://manual.seafile.com/config/seahub_settings_py/" />
+        for all available options.
+      '';
+    };
+  };
+
+  ###### Implementation
+
+  config = mkIf cfg.enable {
+
+    systemd.targets.seafile = {
+      wantedBy = [ "multi-user.target" ];
+      description = "Seafile components";
+    };
+
+    systemd.services.seaf-server = {
+      description = "Seafile server";
+      partOf = [ "seafile.target" ];
+      after = [ "network.target" ];
+      wantedBy = [ "seafile.target" ];
+      serviceConfig = {
+        User = "seafile";
+        Group = "seafile";
+        DynamicUser = true;
+        ProtectHome = true;
+        PrivateUsers = true;
+        PrivateDevices = true;
+        ProtectClock = true;
+        ProtectHostname = true;
+        ProtectProc = "invisible";
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        ProtectKernelLogs = true;
+        ProtectControlGroups = true;
+        RestrictNamespaces = true;
+        RestrictAddressFamilies = [ "AF_UNIX" "AF_INET" ];
+        LockPersonality = true;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+        MemoryDenyWriteExecute = true;
+        SystemCallArchitectures = "native";
+        StateDirectory = "seafile";
+        RuntimeDirectory = "seafile";
+        LogsDirectory = "seafile";
+        Type = "forking";
+        ExecStart = ''
+          ${pkgs.seafile-server}/bin/seaf-server \
+          -F ${centralConfigDir} \
+          -c ${ccnetDir} \
+          -d ${dataDir} \
+          -l /var/log/seafile/server.log \
+          -P /run/seafile/server.pid \
+          -p /run/seafile
+        '';
+        PIDFile = "/run/seafile/server.pid";
+      };
+      preStart = ''
+        mkdir -p ${seafRoot}/{conf,data}
+        mkdir -p ${dataDir}/library-template
```

Comment:
This could be replaced with `StateDirectory = [ "seafile" "seafile/conf" "seafile/data" "seafile/data/library-template" ];`, etc...

Reply:
During my tests I faced an error: systemd cannot create the directory if it already exists. Although I've not been able to pinpoint what was creating the directory.
There should be less `mkdir` now anyway.

Reply:
Any more details on this? `StateDirectory` is pretty powerful, so a failure is unusual.

Reply:
The error log looks like this:
```
server # [    5.433436] systemd[819]: seaf-server.service: Failed to set up special execution directory in /var/lib: File exists
server # [    5.437448] systemd[819]: seaf-server.service: Failed at step STATE_DIRECTORY spawning /nix/store/sl94ch8cwa5f1awlkincp3shjhvpma1z-unit-script-seaf-server-pre-start/bin/seaf-server-pre-start: File exists
server # [    5.442685] systemd[1]: seaf-server.service: Control process exited, code=exited, status=238/STATE_DIRECTORY
```

The [documentation]() states:
> If set, when the unit is started, one or more directories by the specified names will be created (including their parents) below the locations defined in the following table.

But it's unclear to me whether this apply to ExecStartPre as well? I suspected that the preScript was messing with directory structure, but the error is still present after removing it.